### ISSUE TITLE
About the data process

### ISSUE THREAD

**Issue by liyihan5520-wq** [XiaqiongFan/DeepRaman#2]
Now,  I have the pure roman dataset  with dimension of 256,  so  I need to retrain the model. But how should I construct the training.npy and testing.npy

**Comment by zmzhang** [comment]
Hi, thanks for your interest in DeepRaman!

The training script (`training.py`) expects two `.npy` files in the `./data/` directory: **`training_X.npy`** and **`training_Y.npy`**. Here is how to construct them for your 256-dimensional pure Raman spectra:

### 1. `training_X.npy` — Input spectral pairs

This file should be a **4D NumPy array** with shape:

```
(num_pairs, num_per_pair, 2, 256)
```

Based on the `process()` function in [`training.py`](https://github.com/XiaqiongFan/DeepRaman/blob/01a0530d74636421b0f7c1e00bd73e16961f996e/DeepRaman/training.py#L183-L195), the data is reshaped as:

```python
Xtrain = Xtrain.reshape(Xtrain.shape[0]*Xtrain.shape[1], Xtrain.shape[2], Xtrain.shape[3], 1)
```

So the original `training_X.npy` is 4D: `(N, M, 2, spectral_length)`, where:
- **Axis 0 (`N`)**: number of sample groups (e.g., number of mixture scenarios)
- **Axis 1 (`M`)**: number of pairs per group (positive + negative pairs)
- **Axis 2 (`2`)**: the two spectra in each pair — **`[0]` = pure/database spectrum, `[1]` = unknown/mixture spectrum**
- **Axis 3 (`256`)**: your spectral dimension (256 data points in your case)

Each pair consists of a pure component spectrum and a mixture (or another pure) spectrum to be compared.

### 2. `training_Y.npy` — Labels

This should have shape `(N, M)` matching the first two dimensions of `training_X.npy`, where each label is:
- **`1`** — the pair is a **match** (the pure component is present in the unknown spectrum)
- **`0`** — the pair is a **non-match** (the pure component is not present)

The `process()` function reshapes the labels to `(N*M, 1)` and then converts them to one-hot encoding via `tf.keras.utils.to_categorical`.

### 3. Example: how to construct the data

```python
import numpy as np

# Suppose you have:
#   pure_spectra: shape (num_pure, 256) — your pure Raman database
#   mix_spectra:  shape (num_mix, 256) — mixture spectra (or synthetic mixtures)
#   labels:       for each (mix, pure) pair, 1 if pure is in the mix, 0 otherwise

num_mix = mix_spectra.shape[0]
num_pure = pure_spectra.shape[0]

# Build spectral pairs
# Here M = num_pure (for each mixture, pair it with every pure spectrum)
training_X = np.zeros((num_mix, num_pure, 2, 256))
training_Y = np.zeros((num_mix, num_pure))

for i in range(num_mix):
    for j in range(num_pure):
        training_X[i, j, 0, :] = pure_spectra[j, :]    # pure component
        training_X[i, j, 1, :] = mix_spectra[i, :]      # unknown/mixture
        training_Y[i, j] = labels[i, j]                  # 1=match, 0=non-match

np.save('./data/training_X.npy', training_X)
np.save('./data/training_Y.npy', training_Y)
```

### Key notes

1. **No need to normalize beforehand** — the `process()` function in `training.py` already normalizes each spectrum by its max value.
2. **Balance positive and negative pairs** — try to have a roughly balanced number of match (1) and non-match (0) labels for good training performance.
3. **The spectral length (256) works fine** — the model uses `Input((2, None, 1))` with Spatial Pyramid Pooling, so it supports **variable-length spectra**. Your 256-dimensional data will work without modifying the model architecture.
4. **Train/validation split** is handled automatically — `load_data()` uses the first 90% for training and the last 10% for validation.
5. **For testing (`testing.py`)** — you'll need a pure database `.npy` of shape `(num_pure, 256)` and an unknown mixture `.npy` of shape `(num_unknowns, 256)`. Note you'll also need to update the hardcoded spectral dimension `881` on [line 155 of `testing.py`](https://github.com/XiaqiongFan/DeepRaman/blob/01a0530d74636421b0f7c1e00bd73e16961f996e/DeepRaman/testing.py#L155) to `256` to match your data.

Hope this helps! Feel free to open another issue if you run into any problems during retraining.
